src/garage/_environment.py

```diff
@@ -1,6 +1,222 @@
+
+        """
+        self._input_space = action_space
+
+    @observation_space.setter
+    def observation_space(self, observation_space):
+        """Set observation space of the env.
+
+        Args:
+            observation_space (akro.Space): Observation space.
+
+        """
+        self._output_space = observation_space
+
+    @property
+    def max_episode_length(self):
+        """Get max episode steps.
+
+        Returns:
+            int: The maximum number of steps that an episode
+
+        """
+        return self._max_episode_length
+
+
+class EnvStep(
+        collections.namedtuple('EnvStep', [
+            'env_spec', 'observation', 'action', 'reward', 'next_observation',
+            'env_info', 'step_type'
+        ])):
+    # pylint: disable=missing-return-doc, missing-return-type-doc, missing-param-doc, missing-type-doc  # noqa: E501
+    r"""A tuple representing a single step returned by the environment.
+
+    Attributes:
+        env_spec (garage.envs.EnvSpec): Specification for the environment from
+            which this data was sampled.
+        observation (numpy.ndarray): A numpy array of shape :math:`(O^*)`
+            containing the observation for the this time step in the
+            environment. These must conform to
+            :obj:`env_spec.observation_space`.
+            The observation before applying the action.
+            `None` if `step_type` is `StepType.FIRST`, i.e. at the start of a
+            sequence.
+        action (numpy.ndarray): A numpy array of shape :math:`(A^*)`
+            containing the action for the this time step. These must conform
+            to :obj:`env_spec.action_space`.
+            `None` if `step_type` is `StepType.FIRST`, i.e. at the start of a
+            sequence.
+        reward (float): A float representing the reward for taking the action
+            given the observation, at the this time step.
+            `None` if `step_type` is `StepType.FIRST`, i.e. at the start of a
+            sequence.
+        next_observation (numpy.ndarray): A numpy array of shape :math:`(O^*)`
+            containing the observation for the this time step in the
+            environment. These must conform to
+            :obj:`env_spec.observation_space`.
+            The observation after applying the action.
+        env_info (dict): A dict containing environment state information.
+        step_type (StepType): a `StepType` enum value. Can either be
+            StepType.FIRST, StepType.MID, StepType.TERMINAL, StepType.TIMEOUT.
+
+
+    Raises:
+        ValueError: If any of the above attributes do not conform to their
+            prescribed types and shapes.
+
+    """
+
+    def __new__(cls, env_spec, observation, action, reward, next_observation,
+                env_info, step_type):  # noqa: D102
+        # pylint: disable=too-many-branches
+        # observation
+        if not env_spec.observation_space.contains(observation):
```

Comment:
doing all this validation on every step will be really expensive. I think we should leave validation to the batch types (`TimeStepBatch` and `TrajectoryBatch`). Please remove it here and just stick with a simple namedtuple.